proposals/0000-ghc-module-naming.rst

```diff
@@ -0,0 +1,145 @@
+
+
+Timescale
+==========
+The first release of GHC with `ghc-experimental` and `ghc-internals` will be GHC 9.10, which expect to
+release in early 2024.  It would be good to establish naming conventions for modules well before this date.
+
+Example lifecycle
+===================
+
+By way of example, consider the ``HasField`` class, which supports overloaded record fields.
+It is currently defined in ``base:GHC.Records``, which is an odd module to have to import.
+Moreover there is
+more than one GHC proposal that suggest changes to its design (e.g. see `GHC Proposal 158 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0158-record-set-field.rst>`_); it is not nearly as stable as most of ``base``
+
+If ``ghc-experimental`` had existed we would have put it in ``ghc-experimental:Experimental.Records``.
+That would have made it clear that the design of overloaded records still evolving.
+Once the design becomes settled and stable, it could move to ``base``, perhaps in a module like ``Data.Records``.
+
+Other similar examples include
+
+* The tuple proposal of `GHC Proposal 475 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst>`_
+* The `DataToTag CLC proposal <https://github.com/haskell/core-libraries-committee/issues/104>`_ would have been easier to expose through ``ghc-experimental`` in the first instance.
+
+Alternatives
+==============
+* We could dispute Proposal 1: one could imagine deliberately naming modules in ``ghc-experimental`` with the
+  same module name as their eventual expected (by someone) home in ``base``.  The goal would be to reduce impact if and when
+  the module moves from ``ghc-experimental`` to ``base``. For example, we might add ``Data.Tuple`` to ``ghc-experimental`` containing the new type constructors ``Tuple2``, ``Tuple3`` etc that are proposed in `GHC Proposal 475 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst>`_.   However:
+
+  * In the meantime there are two modules both called ``Data.Tuple``.  This is bad.  Which one does ``import Data.Tuple`` import?  (Look at the Cabal file, perhaps?)  How can I import both?  (Package-qualified imports perhaps.) So it will really only help in the case of a brand-new module, not already in ``base``.
+  * It loses the explicit cue, in the source code, given by ``import Experimental.Data.Tuple``.
+
+* We could use ``GHC.*`` for modules in ``ghc-experimental``, and maybe ``GHC.Internals.*`` for module in ``ghc-internals``.  But
+
+  * There are two sorts of GHC-specific-ness to consider:
+
+    * Modules that are part of GHC's implementations
+    * Modules that support a GHC extension, blessed by the GHC Steering Committee
+
+    It is worth distinguishing these: it's confusing if both start with ``GHC.``.
+
+  * It would be a huge upheaval (with impact on users) to rename hundreds of modules in ``ghc-internals``.
```

Comment:
Is this really the case? At the moment `ghc-internals` doesn't yet exist, so nobody depends on it! And the whole point is that most users should not depend on it, they should depend on `base` instead!

So why couldn't we establish the convention that `ghc-internals` uses `GHC.Internal.*` as its preferred module prefix, at least for new modules? We could still move over existing modules from `base` without renaming them if that was practically easier, and then rename them later. Renaming modules in `ghc-internals` should be fairly cheap as users aren't supposed to import them anyway, so the only package affected should be `base`.

Reply:
If we land on `GHC.Internal` and `Foo.Experimental`, I feel like the package should be named `ghc-internal` without the `s`, for consistency

Reply:
> Is this really the case? At the moment ghc-internals doesn't yet exist, so nobody depends on it! And the whole point is that most users should not depend on it, they should depend on base instead!

I anticipate that *many* of the modules in `base` will move to `ghc-internals`, leaving behind a shim module.  But I suppose it is true that when we move, say `GHC.Base` into `ghc-internals`  we could
* Rename it to `GHC.Internal.Base`, in `ghc-internals`
* Leave behind a shim module in `base` that
  * Is called `GHC.Base`
  * Imports `GHC.Internal.Base` and re-exports it all

Hmm.  That's true.   Moreover, because for a long time (possibly forever) we will have modules like `GHC.Base` in `base` (for back-compat reasons) it would be good if the module in `ghc-internals` had a different name.

Reply:
I think it's OK if deprecated modules have non-standard names. If anything it is good; it adds an extra signal saying what ought to be used.

Reply:
> - Leave behind a shim module in base that
>   - Is called `GHC.Base`
>   - Imports `GHC.Internal.Base` and re-exports it all

Yes. Ideally, `GHC.Base` will re-export individual declarations from `GHC.Internal.Base` using an explicit export list (not a `module` export). That way any new definitions added to `GHC.Internal.Base` won't accidentally leak into the `base` API.

So I think it's positively helpful to give the internal modules new names. Another example: we could add module-level `DEPRECATED` pragmas to the `base` shim modules prior to removing them.